kubernetes/istio.go

```diff
@@ -324,49 +325,63 @@ func DestinationRuleHasMTLSEnabled(destinationRule *networking_v1.DestinationRul
 	return false, ""
 }
 
-// ClusterNameFromIstiod attempts to resolve the cluster info of the "home" cluster where kiali is running
-// by inspecting the istiod deployment. Assumes that the istiod deployment is in the same cluster as the kiali pod.
-func ClusterNameFromIstiod(conf config.Config, k8s ClientInterface) (string, error) {
-	// The "cluster_id" is set in an environment variable of
-	// the "istiod" deployment. Let's try to fetch it.
-	var istiodDeployment *appsv1.Deployment
-	if istiodDeploymentName := conf.ExternalServices.Istio.IstiodDeploymentName; istiodDeploymentName != "" {
-		deployment, err := k8s.GetDeployment(conf.IstioNamespace, istiodDeploymentName)
-		if err != nil {
-			return "", err
-		}
+// ClusterNameFromIstiod checks for an istio control plane deployment (istiod) and if found returns the Istio
+// clusterID. Returns "" if no control plane is found or the cluster name can not be determined. Note that
+// this code is limited to the k8s API, the Kiali cache may not yet exist.
+func ClusterNameFromIstiod(conf *config.Config, k8s ClientInterface) (string, error) {
+	// The "cluster_id" is set in an environment variable of the "istiod" deployment. Let's try to fetch it.
+	istiods, err := k8s.GetDeployments("", metav1.ListOptions{LabelSelector: "app=istiod"})
+	if err != nil {
+		return "", err
+	}
 
-		istiodDeployment = deployment
-	} else {
-		istiodDeployments, err := k8s.GetDeployments(conf.IstioNamespace, metav1.ListOptions{LabelSelector: "app=istiod"})
-		if err != nil {
-			return "", err
-		}
+	if len(istiods) == 0 {
+		return "", fmt.Errorf("no valid istiod deployment found in any namespaces for kiali home cluster name discovery. You may need to set kubernetes_config.cluster_name in the Kiali CR")
+	}
 
-		if len(istiodDeployments) == 0 {
-			return "", fmt.Errorf("istiod deployment not found in namespace [%s]", conf.IstioNamespace)
+	// Do our best to filter out any external control plane, because the CLUSTER_ID will be set to
+	// the dataplane cluster, not the hosting cluster (which seems weird, but that's how Istio does it).
+	validIstiods := sliceutil.Filter(istiods, func(istiod v1.Deployment) bool {
+		// a typical namespace name is external-istiod
+		if strings.Contains(strings.ToLower(istiod.Namespace), "external") {
+			log.Tracef("Ignoring likely external controlplane [%s] during Kiali home clustername discovery. Namespace=[%s]", istiod.Name, istiod.Namespace)
+			return false
+		}
+		// there is a decent chance there will be an env var with "EXTERNAL"
+		for _, container := range istiod.Spec.Template.Spec.Containers {
+			for _, v := range container.Env {
+				if strings.Contains(strings.ToLower(v.Name), "external") {
```

Comment:
This should be searching for the full `MANAGES_EXTERNAL` env var (or whatever the name actually is). There may be other env vars that have `external` in them that aren't `MANAGES_EXTERNAL`.

Reply:
I'm not 100% sure from the istio docs, but in the github repo, and in a dump I did in my debugging, I see EXTERNAL_ISTIOD. I'll just limit to this one.